flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -375,6 +378,49 @@ public RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot(
             @Nonnull CheckpointStreamFactory streamFactory,
             @Nonnull CheckpointOptions checkpointOptions)
             throws Exception {
+
+        if (checkpointOptions.getCheckpointType().isSavepoint()) {
+            SnapshotType.SharingFilesStrategy sharingFilesStrategy =
+                    checkpointOptions.getCheckpointType().getSharingFilesStrategy();
+            if (sharingFilesStrategy == SnapshotType.SharingFilesStrategy.NO_SHARING) {
+                // For NO_SHARING native savepoint, trigger delegated one
+                RunnableFuture<SnapshotResult<KeyedStateHandle>> delegatedSnapshotResult =
+                        keyedStateBackend.snapshot(
+                                checkpointId, timestamp, streamFactory, checkpointOptions);
```

Comment:
Yes. 
IIUC, Savepoint should use its own `CheckpointStreamFactory` which is consistent with all state backends as you could see `initializeLocationForSavepoint`, right ?

Reply:
I think using `streamFactory` passed to this method is correct because savepoint should be stored in a separate path.

I have another question: can't `checkpointId` interfere with `materializationId` (used in `initMaterialization`)?
For example, in local recovery paths.

Reply:
Thanks for the reply.
Maybe I missed something.
Currently, `materializationId` is a dummy one while triggering native savepoint.
Do you mean that we should use current real `materializationId` ?

Reply:
I didn't dig into the backend + local recovery code, but I think that would be safer (checkpoint ID received by delegated backend would be monotonically increasing).
WDYT?

Reply:
Thanks for the advice.
I think it's better to let checkpoint id of delegated state backend increase monotonically.
Updated.